lib/path.nix

```diff
@@ -0,0 +1,172 @@
+# Functions for working with paths, see ./path.md
+{ lib }:
+let
+
+  inherit (builtins)
+    isString
+    split
+    match
+    ;
+
+  inherit (lib.lists)
+    length
+    head
+    last
+    genList
+    elemAt
+    ;
+
+  inherit (lib.strings)
+    concatStringsSep
+    substring
+    ;
+
+  inherit (lib.generators)
+    toPretty
+    ;
+
+  pretty = toPretty { multiline = false; };
+
+  # Returns true if the value is a valid subpath string, otherwise throws an error
+  isValidSubpath = value: errorPrefix:
+    if ! isString value then
+      throw "${errorPrefix}:\n    Not a string"
+    else if value == "" then
+      throw "${errorPrefix}:\n    The string is empty"
+    else if substring 0 1 value == "/" then
+      throw "${errorPrefix}:\n    The string is an absolute path because it starts with `/`"
+    else if match "(.*/)?\\.\\.(/.*)?" value != null then
+      # We don't support ".." components, see ./path.md
+      throw "${errorPrefix}:\n    The string contains a `..` component, which is not allowed in subpaths"
+    else true;
+
+  # Splits and normalises a subpath string into its components.
+  # Errors for ".." components and doesn't include "." components
+  splitSubpath = path: errorPrefix:
```

Comment:
We'd run into infinite recursion with `split` though, because we use `inherit (builtins) split` and use it in this function.

Reply:
`splitRelPath` and `joinRelPath`?

Reply:
Oh neat. Also similarly: You can select text from a GitHub comment, then click "Quote Reply" and it will insert only the selected text as a quote.

This is an internal name anyways, we don't need to care about the name that much imo :)

Reply:
> because we use inherit (builtins) split and use it in this function

You can rename the import to `splitString`! Okay, this is getting into a level of detail that suggests the implementation is good enough.

Reply:
Yeah "Quote Reply" is also nice.

> internal

Is it a scandal if no-one knows?

I guess what I was going to say is that "relative" isn't quite right here.

> because we use `inherit (builtins) split`

We don't have to.

I guess the risk is someone might want to expose the function and then suddenly it's not internal anymore. Scandalous!

Reply:
Should I add you @roberth and @fricklerhandwerk as code owners of `lib/path.{nix,md}`? And I'm already a code owner of `lib` so I'll get notifications for future updates to this library